### DISCUSSION TITLE
Github Actions: Provide Branch Name environment variable

### DISCUSSION THREAD

**Discussion by blimmer** [community/community#5251] [Actions]
Currently, there's no easy way to grab the branch name in a GitHub action.

I was previously using `github.head_ref`, but that's only defined in Pull Requests according to the [docs](https://docs.github.com/en/actions/reference/environment-variables#default-environment-variables):

![Screen Shot 2021-08-23 at 11 47 34 AM](https://user-images.githubusercontent.com/630449/130493230-f558f815-12a7-41ef-b4d4-61345895cba5.png)

Because of that, actions that run on my `main` branch don't have `github.head_ref` available.

I would love to see `github.branch_name`, or something similar, available in GH Actions by default to make this easier.

### Most Popular StackOverflow Post

This StackOverflow post https://stackoverflow.com/q/58033366/808678 has lots of activity, showing that many folks are having this issue.

### Okta Blog Post
This [Okta blog post](https://developer.okta.com/blog/2020/05/18/travis-ci-to-github-actions) about migrating from TravisCI to GitHub actions. They call out the migration from `$TRAVIS_BRANCH` as tricky.

<img width="863" alt="Screen Shot 2021-08-23 at 12 45 44 PM" src="https://user-images.githubusercontent.com/630449/130500962-b312abd5-6454-4668-9cd6-1289a786f685.png">

### CodeClimate Test Coverage Issue

CodeClimate mentions in their [documentation](https://docs.codeclimate.com/docs/github-actions-test-coverage) that they can't support GitHub actions very well because of the lack of the branch env var:

<img width="637" alt="Screen Shot 2021-08-23 at 12 57 20 PM" src="https://user-images.githubusercontent.com/630449/130502599-89b3c384-cea2-46c2-b194-26f3d1710761.png">

## Workaround

Currently, I'm using this action https://github.com/marketplace/actions/github-environment-variables-action to inject the variable. However, this is a common enough use-case that I think it should be provided by GitHub Actions natively.

## Other CI Providers

These CI providers all provide a simple env variable that works on PRs and on the base branch builds.

- [Travis CI](https://docs.travis-ci.com/user/environment-variables/#default-environment-variables) - `$TRAVIS_BRANCH`
- [CircleCI](https://circleci.com/docs/2.0/env-vars/#built-in-environment-variables) - `$CIRCLE_BRANCH`
- [Semaphore CI](https://docs.semaphoreci.com/ci-cd-environment/environment-variables/) - `$SEMAPHORE_GIT_BRANCH`
- [GitLab CI](https://docs.gitlab.com/ee/ci/variables/predefined_variables.html) - `$CI_COMMIT_REF_NAME`

**Comment by akumria** [comment]
I wrote a support ticket (see: https://support.github.com/ticket/1380493 ) to ask for a Date for a Date, for this feature.

As an aside, I really wish general feedback was actually labelled with the state of its evaluation (i.e. not started, waiting to hear from others affected, etc.)

Anyway, there are a few reasons why this feature should really be looked at:

1. It is a non-obvious failure mode

When trying to identify a branch, `GITHUB_REF` and `GITHUB_HEAD_REF` seem like obvious candidates. It is easy to create an action that works for one of them but when extending it encounter failure modes. I have now been part of two teams where multiple days were spent learning about, then correcting for this kind of failure.

2. This is easy to fix on the platform (i.e. Github Action) side of things but hard elsewhere.

Scanning through our organisations various repositories reveals (so far) 8 distinct ways that this has been handled by different teams. 

Some utilise github events and extract a common variable name (as the referenced stackoverflow suggests); others conditionalise whole portions of the action based on the event and set different variables.

Other utilise, frankly, baroque means of getting around this platform limitation.

Standardising something here will simplify things for many developers and enhance the readability and debuggability of numerous actions.

**Comment by nicholasgibson2** [comment]
+1 for this functionality

**Comment by grossws** [comment]
They added `GITHUB_REF_NAME`/`github.ref_name` and `GITHUB_REF_TYPE`/`github.ref_type` sometime ago for most common use cases, see https://docs.github.com/en/actions/learn-github-actions/environment-variables

**Reply by youcandanch** [reply]
@grossws that doesn't _entirely_ cover it; while it's great for push events, for pull_request events, you're still getting `<pr_number>/merge`. I get that it's work-around-able by checking event type and then using `head_ref` for PRs, but it does seem like having `branch_name` would be a great quality-of-life improvement for GHA users.

**Reply by grossws** [reply]
Yeah, but still better than always parsing github.ref

**Comment by zulman** [comment]
+1 for this functionality

**Comment by sozercan** [comment]
It seems like easiest workaround is to use:
`SOURCE_BRANCH="${GITHUB_HEAD_REF:-$GITHUB_REF_NAME}"`

**Comment by sr42-dev** [comment]
Bump.

As an alternative, you could also use [this action](https://github.com/tj-actions/branch-names) from the marketplace to get the branch name as a step output. 
I prefer this to using `$GITHUB_HEAD_REF` as it's agnostic of the workflow trigger type.
